pkg/controllers/management/node/controller.go

```diff
@@ -223,71 +222,80 @@ func (m *Lifecycle) getNodePool(nodePoolName string) (*apimgmtv3.NodePool, error
 	return m.nodePoolLister.Get(ns, p)
 }
 
-func (m *Lifecycle) Remove(obj *apimgmtv3.Node) (runtime.Object, error) {
-	if obj.Status.NodeTemplateSpec == nil {
-		if err := m.cleanRKENode(obj); err != nil {
-			return obj, err
+func (m *Lifecycle) Remove(machine *apimgmtv3.Node) (obj runtime.Object, err error) {
+	if machine.Annotations[userNodeRemoveCleanupAnnotation] != "true" {
+		// if a node is provisioned on <v2.5.1, rancher is stopped, and the node is deleted; the user-node-remove finalizer
+		// should just be removed here.
+		if machine, err = m.userNodeRemoveCleanup(machine); err != nil {
```

Comment:
I am slightly concerned that the logic here is checking annotations to decide whether to remove a finalizer. Because we are calling the `userNodeRemoveCleanup` in a `Remove`, the machine's `DeletionTimestamp` will always be not nil, and thus we will constantly be calling `Update` on the object which doesn't make sense to me.

We don't care about the `userNodeRemoveCleanupAnnotation` or finalizer anymore, do we?

Reply:
We only update the object if it's not nil, but if that's confusing I can refactor it.